components/script/dom/bluetooth.rs

```diff
@@ -100,6 +79,46 @@ impl Clone for AllowedBluetoothDevice {
     }
 }
 
+#[derive(HeapSizeOf, JSTraceable)]
+pub struct BluetoothExtraPermissionData {
+    permission_data: DOMRefCell<BluetoothPermissionData>,
+}
+
+impl BluetoothExtraPermissionData {
+    pub fn new() -> BluetoothExtraPermissionData {
+        BluetoothExtraPermissionData {
+            permission_data: DOMRefCell::new(BluetoothPermissionData { allowedDevices: Vec::new() }),
+        }
+    }
+
+    pub fn add_new_allowed_device(&self, allowed_device: AllowedBluetoothDevice) {
+        self.permission_data.borrow_mut().allowedDevices.push(allowed_device);
+    }
+
+    fn get_allowed_devices(&self) -> Vec<AllowedBluetoothDevice> {
+        self.permission_data.borrow().allowedDevices.clone()
```

Comment:
If we return `Ref<Vec<...>>` instead (via Ref::map) then we don't need to clone the array every time.